### PR TITLE
Truncate shifts to the maximum constant bit-width in the constant folding pass.

### PR DESCRIPTION
**Pull request by fruffy**
IR::Constant throws a warning when the provided value overflows. This can lead to throwing unnecessary duplicated warnings in constant folding when working with shifts.

Instead, we just truncate the shifted bit value to the maximum and pass this value. We still have the option to throw a warning with the `warnings` flag.

### CODE REVIEW

File: testdata/p4_16_samples_outputs/gauntlet_various_ops-bmv2.p4-stderr [hunk 1]
```diff
@@ -1,21 +1,18 @@
+gauntlet_various_ops-bmv2.p4(118): [--Wwarn=overflow] warning: 4s12: signed value does not fit in 4 bits
+        h.rshift.d = -4s7 >> 1 >> 1;
+                     ^^^^^^^^^
+gauntlet_various_ops-bmv2.p4(118): [--Wwarn=overflow] warning: 4s14: signed value does not fit in 4 bits
+        h.rshift.d = -4s7 >> 1 >> 1;
+                     ^^^^^^^^^^^^^^
```
Comment (ChrisDodd) [inline testdata/p4_16_samples_outputs/gauntlet_various_ops-bmv2.p4-stderr:6]: Something has gone awry here -- these messages are wrong (and different!)

The result of `-4s7 >> 2` should be `-4s2`

File: frontends/common/constantFolding.cpp [hunk 2]
```diff
@@ -803,16 +803,17 @@ const IR::Node *DoConstantFolding::shift(const IR::Operation_Binary *e) {
...
     auto tb = left->type->to<IR::Type_Bits>();
+    if (e->is<IR::Shl>())
+        value = Util::shift_left(value, shift);
+    else
+        value = Util::shift_right(value, shift);
     if (tb != nullptr) {
         if (((unsigned)tb->width_bits() <= shift) && warnings)
             ::warning(ErrorType::WARN_OVERFLOW, "%1%: Shifting %2%-bit value with %3%", e,
                       tb->width_bits(), shift);
+        value = value & (pow(big_int{2}, tb->width_bits()) - 1);
```
Comment (ChrisDodd) [inline frontends/common/constantFolding.cpp:814]: The result of a shift overflow should be 0 (or -1 for signed right shift of a negative value) according to the spec.

Comment (fruffy) [reply frontends/common/constantFolding.cpp:814]: Previously we deferred to `handleOverflow` which iirc does not have that behavior. If the spec says that the shift is 0 I can also change that.

File: frontends/common/constantFolding.cpp [hunk 3]
```diff
@@ -802,17 +802,22 @@ const IR::Node *DoConstantFolding::shift(const IR::Operation_Binary *e) {
...
-    if (e->is<IR::Shl>())
+    if (e->is<IR::Shl>()) {
         value = Util::shift_left(value, shift);
-    else
+    } else {
         value = Util::shift_right(value, shift);
+    }
+    if (const auto *tb = left->type->to<IR::Type_Bits>()) {
+        // According to the P4 specification, the result of a shift that is larger than the width of
+        // the bit type is 0.
```
Comment (ChrisDodd) [inline frontends/common/constantFolding.cpp:812]: unless it is a negative signed value being shifted right, in which case it should be -1 (sign bit shifted into all bit positions)

File: frontends/common/constantFolding.cpp [hunk 4]
```diff
@@ -802,17 +802,26 @@ const IR::Node *DoConstantFolding::shift(const IR::Operation_Binary *e) {
...
+        if ((static_cast<unsigned>(tb->width_bits()) <= shift)) {
+            if (warnings) {
+                ::warning(ErrorType::WARN_OVERFLOW, "%1%: Shifting %2%-bit value with %3%", e,
+                          tb->width_bits(), shift);
+            }
+            if (tb->isSigned && value < 0) {
```
Comment (ChrisDodd) [inline frontends/common/constantFolding.cpp:818]: need a `e->is<IR::Shr>()` test here too

**Review by ChrisDodd — APPROVED** [review]

**Pull request merged**